comparison.md

```diff
@@ -174,16 +345,17 @@ Both libraries have some sharp edges:
 
 Under AssertJ,
 `assertThat(uniqueIdGenerator.next()).isNotSameAs(uniqueIdGenerator.next())` can
-pass, even though if both `next()` calls return the same value. Under Truth,
-we've chosen a different name for the method (`isNotSameInstanceAs`) to steer
-people away from using it unless they mean to test reference equality.
+pass even if both `next()` calls return the same value. Under Truth, we've
+chosen a different name for the method (`isNotSameInstanceAs`) to steer people
+away from using it unless they mean to test reference equality.
 
 Under Truth, `assertThat(listOfStrings).doesNotContain(integer)` passes, even
-though your test is probably buggy. Under AssertJ, it doesn't compile.
+though your test is probably buggy. Under AssertJ, it doesn't compile. We plan
+to add static analysis to [Error Prone] to catch such bugs.
```

Comment:
@cpovirk I'm curious to know why this something you need static analysis for rather than having the compiler check it for you. Would you mind explaining the rationale behind this? :)

Reply:
It's a similar argument to https://errorprone.info/bugpattern/CollectionIncompatibleType

Plus, when type inference fails on a call that _should_ be permitted, the error messages are ugly:

```
javatests/com/google/common/collect/SetsTest.java:779:
containsExactly(java.util.List<java.lang.Object&java.io.Serializable&java.lang.Comparable<?
extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>...)
in org.junit.contrib.truth.subjects.IterableSubject<capture#106 of ?
extends org.junit.contrib.truth.subjects.CollectionSubject<?,java.util.List<java.lang.Object&java.io.Serializable&java.lang.Comparable<?
extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>,java.util.Collection<java.util.List<java.lang.Object&java.io.Serializable&java.lang.Comparable<?
extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>>>,java.util.List<java.lang.Object&java.io.Serializable&java.lang.Comparable<?
extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>,java.util.Collection<java.util.List<java.lang.Object&java.io.Serializable&java.lang.Comparable<?
extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>>>
cannot be applied to
(java.util.List<java.lang.Object>,java.util.List<java.lang.Object>,java.util.List<java.lang.Object>,java.util.List<java.lang.Object>)
    assertThat(cartesianProduct(x, y)).containsExactly(exp1,
exp2, exp3, exp4);
```

I'm still not 100% convinced that we made the right choice, but that's the theory. I'll note that in the docs. Thanks.

Reply:
Hmm, actually, that's an old message from the days in which `Subject` and its subclasses had type parameters. Let me see what it would look like today.

Reply:
It looks like probably even from an older javac, too.

Reply:
(Back in the day, we didn't have `@SafeVarargs`, either, but that's now a non-issue.)

Reply:
And interestingly, that particular error might not happen with the current javac.

So, while the logic of CollectionIncompatibleType still applies, `E...` has certainly gotten more attractive from all the other changes.

Reply:
No, wait, I messed up my change to `IterableSubject`. Everything I said above is wrong :)

Reply:
OK, trying again: Truth-that-requires-`T` _would_ still fail on the `cartesianProduct` example, even under a new javac. The error is now:

```
core/src/test/java/com/google/common/truth/IterableSubjectTest.java:56: error: method containsExactly in class IterableSubject<E> cannot be applied to given types;
      .containsExactly(exp1, exp2, exp3, exp4);
      ^
  required: List<INT#1>[]
  found:    List<Object>,List<Object>,List<Object>,List<Object>
  reason: varargs mismatch; List<Object> cannot be converted to List<INT#1>
  where E is a type-variable:
    E extends Object declared in class IterableSubject
  where INT#1,INT#2 are intersection types:
    INT#1 extends Object,Serializable,Comparable<? extends INT#2>
    INT#2 extends Object,Serializable,Comparable<?>
```

And, as a digression here into something else relevant to my interests, the error message is just slightly longer if `IterableSubject` has self-type and actual-type parameters:

```
core/src/test/java/com/google/common/truth/IterableSubjectTest.java:56: error: method containsExactly in class IterableSubject<S,A,E> cannot be applied to given types;
      .containsExactly(exp1, exp2, exp3, exp4);
      ^
  required: List<INT#1>[]
  found:    List<Object>,List<Object>,List<Object>,List<Object>
  reason: varargs mismatch; List<Object> cannot be converted to List<INT#1>
  where S,A,E are type-variables:
    S extends IterableSubject<S,A,E> declared in class IterableSubject
    A extends Iterable<E> declared in class IterableSubject
    E extends Object declared in class IterableSubject
  where INT#1,INT#2 are intersection types:
    INT#1 extends Object,Serializable,Comparable<? extends INT#2>
    INT#2 extends Object,Serializable,Comparable<?>
```

Reply:
> It's a similar argument to https://errorprone.info/bugpattern/CollectionIncompatibleType

Ahh, okay. I'm reasonably happy to accept that, then. :)